File: user-service/src/main/java/com/microservice/user/config/SpringSecurityAuditorAware.java
package com.microservice.user.config;

import org.springframework.data.domain.AuditorAware;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
import org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser;
import org.springframework.stereotype.Component;

import java.util.Optional;

/**
 * Class for auditing mongo entities changes.
 */
@Component
public class SpringSecurityAuditorAware implements AuditorAware<String> {

    @Override
    public Optional<String> getCurrentAuditor() {
        return Optional.ofNullable(SecurityContextHolder.getContext().getAuthentication())
            .map(a -> {
                if (a instanceof OAuth2AuthenticationToken) {
                    DefaultOidcUser oidcUser = (DefaultOidcUser) a.getPrincipal();
                    return oidcUser.getEmail();
                } else {
                    return a.getName();
                }
            });
    }
}

File: kotlin-service/src/main/kotlin/com/microservice/kotlin/config/SpringSecurityAuditorAware.kt
package com.microservice.kotlin.config

import org.springframework.data.domain.AuditorAware
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser
import org.springframework.stereotype.Component
import java.util.*

/**
 * Class for auditing mongo entities changes.
 */
@Component
class SpringSecurityAuditorAware : AuditorAware<String> {

    /**
     * Return current logged user or default.
     * @return current user
     */
    override fun getCurrentAuditor(): Optional<String> {
        return Optional.ofNullable(SecurityContextHolder.getContext().authentication)
            .map { a: Authentication ->
                if (a is OAuth2AuthenticationToken) {
                    val oidcUser = a.principal as DefaultOidcUser
                    return@map oidcUser.email
                } else {
                    return@map a.name
                }
            }
    }

}
